### PR TITLE
Emergency mode

### PR DESCRIPTION
**Pull request by byroot**
We talked about that feature a while ago with @gmalette, the idea being to toogle a flag that allows you to trigger deploys in any circumstances.

@graemej just asked for that feature again, so here it is.

The first commit is a small refactoring of the button, the second is of higher importance.

@gmalette @davidcornu mind giving this a look please?

### CODE REVIEW

File: app/helpers/stacks_helper.rb [hunk 1]
```diff
@@ -1,4 +1,16 @@
 module StacksHelper
+  DISABLED_BUTTON = %w(btn--disabled disabled).freeze
```
Comment (davidcornu) [inline app/helpers/stacks_helper.rb]: Is this used anywhere? You're using `classes = %W(btn btn--primary deploy-action #{commit.state})` on line `6`.

Comment (davidcornu) [inline app/helpers/stacks_helper.rb]: Also I'd call this `DISABLED_BUTTON_CLASSES` or something to that effect

File: app/helpers/stacks_helper.rb [hunk 2]
```diff
@@ -26,4 +38,24 @@ def pull_request_link(commit)
   def render_raw_commit_id_link(commit)
     link_to(commit.short_sha, github_commit_url(commit), target: '_blank', class: 'number')
   end
+
+  private
+
+  def deploy_button_disabled?(commit)
```
Comment (davidcornu) [inline app/helpers/stacks_helper.rb]: `can_deploy_commit?` maybe?

File: app/helpers/stacks_helper.rb [hunk 3]
```diff
@@ -26,4 +38,24 @@ def pull_request_link(commit)
...
+    if commit.deployable?
+      return 'Deploy in progress...' if commit.stack.deploying?
+      return 'Deploy'
+    end
+
+    return 'CI Pending...' if commit.pending?
+    return 'CI Failure' if commit.failure?
+    return 'CI Error' if commit.error?
+    'Not Run'
+  end
```
Comment (davidcornu) [inline app/helpers/stacks_helper.rb]: Might read slightly better as a `case`. Up to you.

```ruby
case
when commit.stack.locked? then "Locked"
when commit.stack.deploying? then "Deploy in progress..."
when commit.stack.deployable? then "Deploy"
when commit.pending? then "CI Pending..."
when commit.failure? then "CI Failure"
when commit.error? then "CI Error"
else "Not Run"
end
```

**Comment by davidcornu** [comment]
Minor stuff. Haven't seen UI. Other than that :ship:

**Comment by gmalette** [comment]
> Haven't seen UI.

screen shot 2015-06-15 at 11 37 30 am

**Comment by byroot** [comment]
Here's the UI:

capture d ecran 2015-06-15 a 17 37 14

capture d ecran 2015-06-15 a 17 37 21

**Comment by davidcornu** [comment]
Lol thanks guys. UI LGTM

File: app/views/stacks/show.html.erb [hunk 4]
```diff
@@ -37,6 +37,14 @@
   <section>
     <header class="section-header">
       <h2>Undeployed Commits</h2>
+
+      <div class="commit-list-actions">
+        <% if params[:force] %>
+          <%= link_to 'Exit emergency mode', stack_path(@stack) %>
+        <% else %>
+          <%= link_to 'Enable emergency mode', stack_path(@stack, force: 1) %>
```
Comment (gmalette) [inline app/views/stacks/show.html.erb]: I find `force` ambiguous, why not use `emergency`?

**Comment by gmalette** [comment]
:+1:

File: app/views/stacks/show.html.erb (hunk 4 above)
Comment (byroot) [inline app/views/stacks/show.html.erb]: It was to mirror the deploy API `force` parameter, but you're probably right.

**Comment by byroot** [comment]
Thanks for the review guys!

File: app/helpers/stacks_helper.rb (hunk 2 above)
Comment (rafaelfranca) [inline app/helpers/stacks_helper.rb]: I think `deploy_button_disabled?` fits more with this case. Usually I use `can_\w+?` something when we method receiver is an actor, like a user, in this case the receiver is the helper and it is related to Stack that is not the actor of the action.

File: app/helpers/stacks_helper.rb (hunk 3 above)
Comment (rafaelfranca) [inline app/helpers/stacks_helper.rb]: :+1:

File: app/helpers/stacks_helper.rb [hunk 5]
```diff
@@ -1,4 +1,14 @@
 module StacksHelper
+  def deploy_button(commit)
```
Comment (rafaelfranca) [inline app/helpers/stacks_helper.rb]: Maybe some tests for this helper?

**Comment by dwradcliffe** [comment]
:+1: nice feature

File: app/helpers/stacks_helper.rb (hunk 5 above)
Comment (byroot) [inline app/helpers/stacks_helper.rb]: Indeed. That's a bad habit I have to almost never test views and helpers /o\

File: app/helpers/stacks_helper.rb (hunk 3 above)
Comment (byroot) [inline app/helpers/stacks_helper.rb]: Sorry I fixed it, but look like I forgot to commit the change. I just pushed it on master.

**Pull request merged**